### ISSUE TITLE
Entirely remove AbstractComponent

### ISSUE THREAD

**Issue by nik9000** [elastic/elasticsearch#34488]
`AbstractComponent` is trouble because its name implies that
*everything* should extend from it. It *is* useful, but maybe too
broadly useful. The things it offers access too, the `Settings` instance
for the entire server and a logger are nice to have around, but not
really needed *everywhere*.

So we should remove it and everything that needs a logger or a deprecation logger or a settings instance should just declare that it needs it. That way tests don't end up making a `Settings` instance "just because". That kind of thing makes test much harder to read.

See [this](https://github.com/elastic/elasticsearch/issues/34488#issuecomment-434902624) comment for the part of this that is a "good first issue".

**Comment by nik9000** [comment]
Once I merge #35140 I think this will be another one of those "good first issue", "help wanted" things where we folks can grab a couple of packages at the time and remove `extends AbstractComponent` and add a `private static final Logger logger` member.

**Comment by nik9000** [comment]
OK! I've merged #34691 so a good chunk of this is now "good first issue" territory. In particular, replaceing `extends AbstractComponent` with `private static final Logger logger = LogManager.getLogger(ThisClass.class);`. I see 141 `extends AbstractComponent`s so I figure this'd be good to break up. Something like:
* [ ] server
* [ ] x-pack
* [ ] everywhere else

That way we don't get huge PRs. Which are not great for first PRs because they can have many merge conflicts.

**Comment by nik9000** [comment]
I'd like to think a little more about TransportAction and BaseRestHandler. Those are very squarely in the plugin API and I wonder if in 6.x they should keep to `protected final Logger logger = LogManager.getLogger(getClass());` on them. I've made so many breaking change for plugin developer in my quest to remove all of this ceremony I'm feeling a bit bad about it.

**Comment by jklancic** [comment]
Can I take this issue? I would like to contribute.

**Comment by nik9000** [comment]
@jkelastic would you like to remove `extends AbstractComponent` from the non-server, non-xpack portions of the code base? That hasn't been claimed.

**Comment by jklancic** [comment]
Sure, I would love that @nik9000

**Comment by nik9000** [comment]
> Sure, I would love that @nik9000

I've added your name to my tracking comment. Enjoy! Thanks!

**Comment by jklancic** [comment]
I managed to get all remaining references removed. But I have an issue running the job `./gradlew check`. I keep getting

```
> Task :distribution:bwc:staged:buildBwcVersion FAILED

 [6.5.0] ERROR: JAVA_HOME is set to an invalid directory: null
 [6.5.0] 
 [6.5.0] Please set the JAVA_HOME variable in your environment to match the
 [6.5.0] location of your Java installation.
 [6.5.0] 

FAILURE: Build failed with an exception.
```
I did not change anything inside the distribution modules. I only touched modules: plugins, test, server (there was a missing class)

**Comment by nik9000** [comment]
@jkakavas, `Task :distribution:bwc:staged:buildBwcVersion FAILED` is usually caused by us bumping the version of Elasticsearch in an old branch. You'd need to merge master to get it.

**Comment by siddu244** [comment]
Hey guys, iam new to open source
I am eager to contribute in elasticsearch, but facing issue in understanding the basic codebase of elastic search. Can someone provide me a documentaion of code base structure or help in understanding a basic code flow of elastic search it would be too helpful

**Comment by jkakavas** [comment]
> @jkakavas, Task :distribution:bwc:staged:buildBwcVersion FAILED is usually caused by us bumping the version of Elasticsearch in an old branch. You'd need to merge master to get it.

ping @jklancic for visibility. (_too many `@jk*` usernames in one issue_ )

**Comment by nik9000** [comment]
Thanks @jkakavas! Sorry for that. I'm too quick on the keyboard sometimes. Every time I see github folks I ask them to implement nicer type-ahead features for `@` in issues and PRs. It isn't an excuse, but it'd be nice!

@siddu244 I've actually just removed `good first issue` and `help wanted` from this issue as I found a victim on the Elasticsearch team and the remaining steps, while not hard, do touch a *ton* of code. As far as a general tour, I gave a [presentation](https://www.elastic.co/elasticon/conf/2016/sf/contributing-to-elasticsearch-how-to-get-started) a while back that might be nice. But I expect it isn't as useful as finding something small and working on it. We don't have a current "big map" of the code base that I can point you to.
